service/http/topic_test.go

```diff
@@ -49,18 +49,117 @@ func testErrorTopicFunc(ctx context.Context, e *common.TopicEvent) (retry bool,
+		s := newServer("", nil)
+		sub := &common.Subscription{
+			PubsubName: "messages",
+			Topic:      "test",
+			Route:      "/",
+			Metadata:   map[string]string{},
+		}
+		err := s.AddBulkTopicEventHandler(sub, nil, 10, 1000)
+		require.Error(t, err, "expected error adding event handler")
+	})
+}
+
+func TestEventHandler(t *testing.T) {
+	data := `{
+		"specversion" : "1.0",
+		"type" : "com.github.pull.create",
+		"source" : "https://github.com/cloudevents/spec/pull",
+		"subject" : "123",
+		"id" : "A234-1234-1234",
+		"time" : "2018-04-05T17:31:00Z",
+		"comexampleextension1" : "value",
+		"comexampleothervalue" : 5,
+		"datacontenttype" : "application/json",
+		"data" : "eyJtZXNzYWdlIjoiaGVsbG8ifQ=="
+	}`
+
 	s := newServer("", nil)
+
 	sub := &common.Subscription{
 		PubsubName: "messages",
 		Topic:      "test",
 		Route:      "/",
 		Metadata:   map[string]string{},
 	}
-	err := s.AddTopicEventHandler(sub, nil)
-	require.Errorf(t, err, "expected error adding event handler")
+	err := s.AddTopicEventHandler(sub, testTopicFunc)
+	require.NoError(t, err, "error adding event handler")
+
+	sub2 := &common.Subscription{
+		PubsubName: "messages",
+		Topic:      "errors",
+		Route:      "/errors",
+		Metadata:   map[string]string{},
+	}
+	err = s.AddTopicEventHandler(sub2, testErrorTopicFunc)
+	require.NoError(t, err, "error adding error event handler")
+
+	sub3 := &common.Subscription{
+		PubsubName: "messages",
+		Topic:      "test",
+		Route:      "/other",
+		Match:      `event.type == "other"`,
+		Priority:   1,
+	}
+	err = s.AddTopicEventHandler(sub3, testTopicFunc)
+	require.NoError(t, err, "error adding error event handler")
+
+	s.registerBaseHandler()
+
+	req, err := http.NewRequest(http.MethodGet, "/dapr/subscribe", nil)
+	require.NoErrorf(t, err, "error creating request: %s", data)
+	req.Header.Set("Accept", "application/json")
+	rr := httptest.NewRecorder()
+	s.mux.ServeHTTP(rr, req)
+	resp := rr.Result()
+	defer resp.Body.Close()
+	payload, err := io.ReadAll(resp.Body)
+	require.NoError(t, err, "error reading response")
+	var subs []internal.TopicSubscription
+	require.NoError(t, json.Unmarshal(payload, &subs), "could not decode subscribe response")
+
+	sort.Slice(subs, func(i, j int) bool {
+		less := strings.Compare(subs[i].PubsubName, subs[j].PubsubName)
+		if less != 0 {
+			return less < 0
+		}
+		return strings.Compare(subs[i].Topic, subs[j].Topic) <= 0
+	})
+
+	if assert.Len(t, subs, 2, "unexpected subscription count") {
+		assert.Equal(t, "messages", subs[0].PubsubName)
+		assert.Equal(t, "errors", subs[0].Topic)
+
+		assert.Equal(t, "messages", subs[1].PubsubName)
+		assert.Equal(t, "test", subs[1].Topic)
+		assert.Equal(t, "", subs[1].Route)
+		assert.Equal(t, "/", subs[1].Routes.Default)
+		if assert.Len(t, subs[1].Routes.Rules, 1, "unexpected rules count") {
+			assert.Equal(t, `event.type == "other"`, subs[1].Routes.Rules[0].Match)
+			assert.Equal(t, "/other", subs[1].Routes.Rules[0].Path)
+		}
+	}
```

Comment:
This is quite confusing to read - would it be more idiomatic to create a known map and compare this to the subscriptions?